[BE] feat: logging 추가 #204

gabean13 opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->

- develop이랑 merge는 머지 직전에 하겠습니다

Logging Aspect 클래스와 log4j2.xml 클래스 위주로 보시면 될 것 같습니다.

현재 로깅 전략
- log.txt (info, warn, error, fatal) 실시간 운영상황 확인용, 1일 마다 rolling, 7개쌓이면 오래된 것 부터 삭제
- error-log.json (error, fatal) 에러 추적 및 분석용, 1일마다 rolling, 20개 쌓이면 오래된 것 부터 삭제

- getMapping, postMapping, deleteMapping, putMapping, patchMapping 으로 들어오고, return할 때 로깅 (request, reponse)
- RestControllerAdvice, ControllerAdvice return 시 출력 (error)

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

- 콘솔에 출력되는 형식 가독성 괜찮은지, 부족한 정보는 없는지 확인 부탁드립니다. (request, response, error)

- json이 현재 message안에 우리가 출력되는 값이 저장되는데 이런식으로 저장되어도 cloudWatch에서 인식 가능한지 해봐야할것같아요.

- 외부 API를 위한 Annotaion을 만들어서 해당 Annotation 메서드 동작 시 로깅되도록 하는 거 어떤가요? 아직 추가는 못했슴둥.....

- 민감 정보 마스킹 어떻게 하면 좋을지. 현재는 message로 출력되기 전에 전부 objectMapper를 거치지 때문에 json read를 막아놓은 상태인데, toString()하면 다 찍혀서 반쪽자리 대처임..


## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->

close #204

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * Spring Boot 백엔드에 Log4j2 기반의 구조적 로깅 시스템이 도입되었습니다.
  * HTTP 요청/응답 및 예외 상황에 대해 민감 정보 마스킹과 함께 JSON 포맷의 상세 로그가 기록됩니다.
  * 에러 발생 시 별도의 에러 로그 파일이 생성되며, 일자별로 롤링 및 보관됩니다.
  * 요청 및 응답 정보를 포함한 상세 로그와 예외 정보를 구조화하여 기록하는 로깅 컴포넌트가 추가되었습니다.
  * JWT 인증 필터에 에러 로깅 기능이 강화되었습니다.

* **환경설정**
  * 개발 환경에서 Log4j2 설정 파일이 적용되도록 설정이 추가되었습니다.
  * 기존 Spring Boot 기본 로깅이 Log4j2로 대체되었습니다.
  * 로그 출력 포맷과 롤링 정책을 정의하는 Log4j2 XML 설정 파일이 추가되었습니다.

* **버그 수정 및 개선**
  * 글로벌 예외 처리기에서 중복 로깅이 제거되어, 로그 일관성이 향상되었습니다.

* **기타**
  * 불필요한 로그 디렉터리가 버전 관리에서 제외됩니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gabean13 committed:
[BE] chore: .gitignore /log 추가 #204

- 로컬 log 파일 커밋 방지

gabean13 committed:
[BE] chore: log4j2 의존성 추가 및 logback 의존성 제거 #204

gabean13 committed:
[BE] chore: log4j2 configuration 파일 작성 #204

gabean13 committed:
[BE] feat: Logging Aspect 추가 #204

- request, response, error LoggingMessage 추가
- Json으로 변환해주는 class 추가

gabean13 committed:
[BE] refactor: json 변환 시 중요 정보 포함되지 않도록 수정 #204

gabean13 committed:
[BE] chore: error logger 분리 #204

gabean13 committed:
[BE] refactor: objectMapper, DefaultPrettyPrinter bean으로 등록 #204

gabean13 committed:
[BE] feat: error, warn을 분리하여 logging #204

- null 체크 추가
- 메서드 리팩토링

gabean13 committed:
[BE] feat: filter 로깅 추가 #204

gabean13 commented:
@aicoderabbit 코드레빗 리뷰 다시해줘

gabean13 committed:
[BE] refactor: 로그 레벨 네이밍 수정 #204

gabean13 committed:
[BE] feat: 중요 정보 마스킹 기능 추가 #204

JO-eusan commented on backend/.gitignore:
```diff
@@ -33,6 +33,9 @@ out/
 /nbdist/
 /.nb-gradle/
 
+#log
```
사소하긴 하지만, 위에 주석이랑 형식을 맞춰서 `### Logging ###`으로 명확히 명시하고 코드 레빗의 조언처럼 디렉터리를 `/log/`이렇게 하는거 어떨까요?

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/infrastructure/jwt/FilterErrorLogger.java:
```diff
@@ -0,0 +1,39 @@
+
+        log.warn(jsonMaskingPrettyFormatter.convertToPrettyJson(errorLog));
+    }
+}
```
마지막 개행 추가가 빠진 것 같아요!

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/infrastructure/jwt/FilterErrorLogger.java:
```diff
@@ -0,0 +1,39 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class FilterErrorLogger {
```
제가 생각하기에는 이 클래스는 `jwt`와 직접적인 관계가 없다고 생각합니다. 사용하는 곳도 `auth/config`이구요.
따라서 `auth`패키지 안에 `/config`와 `/log`로 나누어서 패키지를 구성하는건 어떨까요?

gabean13 replied:
흠 그러면 차라리 /common/log 안으로 이동시킬까요?

JO-eusan replied:
그것도 좋은거 같아용

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/infrastructure/jwt/FilterErrorLogger.java:
```diff
@@ -0,0 +1,39 @@
+
+    private final JsonMaskingPrettyFormatter jsonMaskingPrettyFormatter;
+
+    public void log(HttpServletRequest request, ProblemDetail problemDetail) {
```
외부에서 봤을 때, 로그 레벨이 추측되지 않아 메서드를 들여다봐야하는 불편함이 있는 것 같습니다.
메서드 이름을 `log` -> `logWarn`으로 변경하는건 어떨까요?

gabean13 replied:
인정합니두

Byesol commented on backend/src/main/java/com/onair/hearit/common/log/LoggingAspect.java:
```diff
@@ -0,0 +1,188 @@
+                });
+    }
+
+    @AfterReturning(value = "exceptionHandler()", returning = "problemDetail")
```
오류 로깅을 위해 위처럼 exceptionHandler에 대해 AfterReturning하는 방식 도 좋지만 @AfterThrowing을 사용하는 건 어떨까요?

현재는로깅 로직이 exceptionHandler 에 의존하기 때문에  만일 누군가 예외 객체를 누락시킨  exceptionHandler를 추가하면 로깅이 누락될 수도 있지 않을까요?

따라서 afterthrowing 을 쓰고, 400번대와 500번대 구분을 위해서 instanceof를 통해 custom 오류와 구분하는 건 어떨까요?(꼭 필요 x, 제안)

gabean13 replied:
근데 그러면 ProblemDetail을 못찍어요ㅠㅠ
아니면 logging 하는 족에서 다시 생성해줘야하는데 그러면 exceptionHandler에서 생성하는 코드랑 서로 동기화해줘야하는 문제가 있어요 그리고 지금 GlobalExceptionHandler에서 Exception으로 커스텀 예외로 잡지 못하는 예외 잡고 있지않나욤..?
제가 놓친부분이 있으면 알려주세욥

Byesol commented on backend/src/main/java/com/onair/hearit/common/log/message/JsonMaskingPrettyFormatter.java:
```diff
@@ -0,0 +1,78 @@
+
+@Component
+@RequiredArgsConstructor
+public class JsonMaskingPrettyFormatter {
```
클래스명과 내부 내용을 보면 Json 포매팅과 masking, 두 가지 역할을 하고 있다고 느껴지는데 역할이 많은 것 같아,
아예 Json 포매팅하는 클래스와 Masking 하는 클래스를 분리하는 건 어떨까요?

gabean13 replied:
맞아요  근데 Object를 처리해서 Json으로 만든다 를 하는 클래스로서 추가적인 분리를 한다고 가독성이 크게 개선된다고 느끼진 못했습니다.. 
그리고 마스킹 하는 부분은 크게 리팩토링이 필요한 부분이라 일단 이렇게 하는거 어떻게 생각하시나요??

Byesol replied:
인정합니다.  approve 할께욥!

Byesol commented on backend/src/main/java/com/onair/hearit/common/log/message/JsonMaskingPrettyFormatter.java:
```diff
@@ -0,0 +1,78 @@
+     * @param visible
+     * @return value의 앞 뒤 1글자 제외하고 마스킹 된 String을 반환한다.
+     */
+    private static String maskEdge(String value, int visible) {
```
(순수궁금) 앞뒤 1글자는 남기는 이유가 뭔가요??

gabean13 replied:
최소한의 구분을 하려고 했습니당. 

찾아봤을 때, 업무상 로그를 추적하고, 구별하기 위해서 아이디 같은 거 완전 제거보다 개인정보를 침해하지 않는 선에서 비식별(특정하지 않을 정도)로 하는게 좋다고 하더라구요

Byesol commented on backend/src/main/java/com/onair/hearit/common/log/LoggingAspect.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Pointcut("getMapping() || postMapping() || deleteMapping() || putMapping() || patchMapping()")
+    public void allMapping() {
+    }
```
제가 강의실에서 대화를 놓친 거일 수도 있는 거 같은데 

저희 '서비스'레이어 쪽도 로깅하기로 하지 않았나요?

gabean13 replied:
맞아요 근데 저희 또 얘기할 때 service나 외부 api는 일단 log.info로 필요한 부분만 찍자고 말했었어요!
이 부분도 나중에 편리하게 하려면 커스텀 어노테이션으로 찍을 수 있게 해야할듯요

Byesol replied:
아맞다. 말하니까 기억이 나네영. 감사합니당

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/config/LogConfig.java:
```diff
@@ -0,0 +1,28 @@
+        printer.indentArraysWith(new DefaultIndenter("  ", System.lineSeparator()));
+        return printer;
+    }
+}
```
한 줄 띄어야할 것 같습니당

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/message/dto/RequestLog.java:
```diff
@@ -0,0 +1,29 @@
+                ));
+        return new RequestLog("REQUEST", timestamp, requestInfo, parameters, body);
+    }
+}
```
마지막 개행이 빠진 것 같습니다.

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/message/dto/ResponseLog.java:
```diff
@@ -0,0 +1,26 @@
+                responseEntity,
+                timeTakenMs);
+    }
+}
```
여기도요!

Byesol commented on backend/src/main/java/com/onair/hearit/common/log/message/JsonMaskingPrettyFormatter.java:
```diff
@@ -0,0 +1,78 @@
+    private static final Map<String, Function<String, String>> MASKING_RULES = new HashMap<>() {{
+        put("password", v -> "*****");
+        put("localId", v -> maskEdge(v, 1));
+        put("accessToken", v -> "*****");
```
지금 이 방식으로 하면, 직접 필드명이 변경되거나 하면, MASKING RULES를 찾아 고쳐야 할 것 같은데

마스킹 부분도 '외부 클라이언트 로깅 관련해 커스텀 어노테이션이라는 아이디어를 통해 구현한 것처럼 마스킹을 위해서도 마스킹이 필요한 클래스의 각필드마다 커스텀 어노테이션을 붙여 해결하는 건 어떨까요?


그리고 마찬가지로 논의하고 싶은 내용

`민감 정보 마스킹 어떻게 하면 좋을지. 현재는 message로 출력되기 전에 전부 objectMapper를 거치지 때문에 json read를 막아놓은 상태인데, toString()하면 다 찍혀서 반쪽자리 대처임..`
의ㅣ 경우 마스킹 원하는 클래스에서 toString을 재정의하는 걸로 해결할 순 없나요? 어떤 문제가 있나요?

yuyoungrhee replied:
여기 제의견..

마스킹 원하는 클래스에서 toString()을 재정의하는 것은 책임구조상 문제가 있을 것 같아요. 로깅은 외부책임인데, dto가 너무 자아가 생기는 느낌 ?

그리고 toString()의 경우 디버깅 등 로깅 말고도 사실상 더 다양한 용도로 쓰일 수 있는데 디버깅하다가 마스킹으로 나오는 등의 문제가 있을듯 합니다..

gabean13 replied:
저도 커스텀 어노테이션 붙혀서 하는 방법이 제일 베스트일 것 같긴한데, 일단 빠르게 구현하는게 목적이라서 일단 이렇게 했어요...
필드명이 변경되면 문자열 찾아서 고쳐야하는 점이 젤 큰 문제인 것 같아요.

이 부분은 빠르게 로깅 운영환경에서 찍히게 지금거 머지시키고 바로 이슈파서 추가하면 좋을 것 같아요! 

그리고 마스킹원하는 클래스를 건드리는게 좀 혹시 모르는 부작용이 있을 것 같아서 지양하려고요 
로깅은 현재 구현 로직에 영향을 미치면 안되는데, toString이나 @JsonSerializiton, @JsonIgnore 같은 어노테이션을 달면 구현 코드나 디버깅에서 의도대로 동작하지 않을 수도 있을 것 같아욤

Byesol commented on backend/src/main/java/com/onair/hearit/common/log/message/dto/ErrorLog.java:
```diff
@@ -0,0 +1,53 @@
+        HttpStatus httpStatus,
+        ErrorDetail errorDetail
+) {
+
```
저희 record 쪽 컨벤션을 따르면 이 dto를 포함해서 다른 dto들까지 처음한 줄 공백 제거하는 게 좋아보여요!

JO-eusan commented on backend/src/main/resources/log4j2/log4j2.xml:
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="WARN">
```
이 설정이 로그 레벨이 `WARN` 이상인 로그만 콘솔에 찍는 거라고 학습했는데, 이건 로그 파일을 만들 때 발생하는 경고와 에러를 콘솔에 찍는다는 의미인걸까요?
학습하다가 의문이 들어서 질문드려요!

gabean13 replied:
이거는 이 log4j2 초기화할때 찍히는 로그 레벨을 WARN으로 설정한다는 거에용!!
그래서 뭐 설정 로딩하거나, 파일 읽을 때 에러나면 찍혀욤

Byesol reviewed (commented):
정말 고생 많으셨어요!!

JO-eusan reviewed (changes requested):
가콩! 디테일한 부분까지 많이 고민해주시고 공유해주신 것 같아요.
궁금한 점에 대해서 커멘트 남겼습니다. 확인 부탁드려요!

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/log/LoggingAspect.java:
```diff
@@ -0,0 +1,188 @@
+                .orElseGet(ErrorDetail::emptyErrorDetail);
+    }
+
+    private void logServerError(RequestInfo requestInfo, HttpStatus httpStatus,
```
`logServerError`와 `logClientError`을 만들 때, `LocalDateTime.now`나 `RequestInfo`의 내용을 가지고 오는 부분은 꼭 Aspect 클래스 내부에서 처리를 해야하나라는 생각이 드네용.
DTO 내부적으로 처리를 하고 errorType 별로 생성하는 정적팩토리메서드를 만들어서 이 클래스 내부에서는 ErrorLog.error, ErrorLog.warn 이런식으로 만든다면 외부에서 가독성이 더 좋지 않을까요?
가콩의 생각은 어떠신가요.

gabean13 replied:
인정 저는 그래서 logType 이넘까지 만들었었음 info, warn, error 근데 하다가 문자열 하나 이넘으로 바꾼거말고는 의미가 없는 것 같아서 삭제했는데 다시 해볼게요 이넘이 확실히 나을것 같긴해요

gabean13 replied:
LocalDateTime.now()는 LoggingAspect에서 하고 넘기는게 낫다고 생각한게 
타임 스탬프 찍히는 시점이 정확히 언젠지 아는게 중요하다고 생각했어요. 

그리고 LogLevel이라는 enum이 이미 springboot logging에 정의되어 있어서 해당 클래스를 사용했습니다.
그래서 errorType별로 정팩매를 호출하도록 했어요 이때, new로 생성하기를 원치 않기 때문에 dto를 class로 바꾸고 생성자 private으로 선언했습니다.

JO-eusan reviewed (commented):
추가로 한가지 제안드릴께 있어서 커멘트 남겨요!

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/common/log/message/dto/ResponseLog.java:
```diff
@@ -0,0 +1,26 @@
+        long timeTakenMs
+) {
+
+    public static ResponseLog of(
```
[로깅과 관련 없음]

그냥 갑자기 드는 생각인데 new를 안 쓰고 정펙메만 열어둘 거면 record가 아니라 class가 맞는 걸지도... 저번에 관련해서 잠깐 얘기했는데 dto는 걍 넘어가자.. 하고 넘어갔었걸랑요
근데 로깅 들어오면서 무조건 정펙메를 쓰는 게 중요해지니까 그런 생각이 더 드네요

gabean13 replied:
인정 저도 이거 그래서 레코드로 바꿧다가 다시 클래스로 바꿧다가 다시 레코드로 바꿨거든요? 
걍 클래스로 바꿀까여

JO-eusan replied:
오 맞네요 정팩메만 쓰도록 유도하게 class로 변경하고 생성자를 막는게 좋아보여요

gabean13 replied:
클래스로 바꾸고 정팩매만 사용하도록 생성자 private으로 선언했습니다

yuyoungrhee reviewed (dismissed):
로깅 어 렵 다... 믿을게요.. 막이래

멍PT와 함께 공부해봤는데, 현재 당장 운영환경에서 돌렸을 때 문제가될만한 부분은 없고, 추후 적용해볼만한 점들만 조금 있다고 해서 저는 Approve할게요!

개선해볼만한 점
1. 클라이언트 X-Request-ID 수용 옵션 추가 - 나중에 하면 좋을듯
이거는 안드랑 얘기해봐야됨요 안드에서 헤더에 추가해서 주는 거예요 조금 더 길게 하나의 요청 흐름을 가져가기 위해서
    - 목적: 프론트–백 간의 요청 상호 추적
	    - 안드 앱이 서버에 요청 → 서버에서 내부 서비스, DB, S3 등 여러 단계 거쳐 응답
	    - 이 전체 흐름을 “하나의 요청”으로 연결해서 추적해야 함
	    - 서버 로그에는 MDC 기반 ID가 찍히고, 클라이언트도 X-Request-ID로 해당 요청의 ID를 알고 있으면, 장애 발생 시 협업 및 추적이 명확

2. Filter 단 로깅 추가
지금은 Spring Security의 Filter에서 예외가 발생할 가능성이 있다는 사실을 알고 직접 대응하고 있는데, 실제로 Controller까지 안 꽂히고 Filter에서 발생하는 예외도 누락되지 않도록 도움
    - OncePerRequestFilter 또는 ErrorController에서 구현 가능하다고 하네요

Byesol reviewed (dismissed):
approve 할께영!

gabean13 committed:
[BE] refactor: dto class로 변경 #204

- 기본 생성자 사용못하도록 class로 변경
- 코드 정렬
- FilterErrorLogger 패키지 이동

gabean13 committed:
[BE] chore: .gitignore 수정 #204

gabean13 committed:
[BE] refactor: Optional 매개변수로 넘기지 않도록 수정 #204

gabean13 committed:
[BE] Merge branch 'develop-be' into feat/#204-logging #204

# Conflicts:
#	backend/src/main/java/com/onair/hearit/common/exception/GlobalExceptionHandler.java
#	backend/src/main/resources/application-dev.properties

gabean13 committed:
[BE] Merge branch 'develop-be' into feat/#204-logging #204

Byesol reviewed (approved):
확인되어 approve 합니다!

yuyoungrhee reviewed (approved):
ㄱㅏ콩 너무너무너무너무너무수고많으셨습니다~~~ 갓콩이셔
Approve할게욧

JO-eusan reviewed (approved):
가콩 Approve & Merge 할게요!
